### ISSUE TITLE
[D8][PS] Allow configuration storage in alternative locations (for example: the database).

### ISSUE THREAD

**Issue by sentaidigital** [backdrop/backdrop-issues#2277]
There is one storage class available for `Config` to use: `ConfigFileStorage`.  I'd like the option of using a `ConfigDatabaseStorage` class that would save the same data, but in the database instead of in flat files.

I can think of the following reasons why:
- Makes it easier to synchronize config changes among multiple web servers in a load balanced environment.
- Makes it easier to capture the entire state of an environment for debugging elsewhere.

PR: https://github.com/backdrop/backdrop/pull/4453
~Replaces: https://github.com/backdrop/backdrop/pull/1613~

--- 
Summary:

This is a feature request for an option that will likely only be used by a minority of sites running Backdrop. As you can see in the discussion below, there are other reasons it might still be a candidate for core inclusion. This option would be disabled by default, so there would be no change for existing sites, or new installs.

Here is a list of pros and cons for PMC review:

Pros
* It's very hard to add this kind of feature in contrib.
* some load-balanced sites may choose use this setting.
* similarity to Drupal 8 / Drupal 7 / WordPress (w/ config in db)
* one less data source for devs to download.
* discourages manual editing of the active configuration (security by obscurity).
* Hosting platforms that were built for config-in-db projects could be used in the way they were intended (config + content together)
* Security improvement: databases tend to be behind firewalls and only accessible by the webheads, which protects configuration better than an easily misconfigured web server directive protecting one directory in an otherwise published files directory.
* possible performance gain when feature IS in use (needs benchmarks)

Cons:
* does not benefit at least 80% of sites using Backdrop.
* possible performance loss when feature IS in use (needs benchmarks)
* possible performance impact when feature is NOT in use (needs benchmarks)
* increased complexity / more code to support (slightly).
* database-storage is less visible than files, making Backdrop harder to learn.
* Hosting platforms that were built for config-in-db projects would no longer benefit from the separation of config and content
* We would need to do all automated tests w/ both file and db configs to avoid any possible exceptions.

**Comment by sentaidigital** [comment]
Getting Backdrop to store configuration in the database required the following changes:

* Adding a `ConfigDatabaseStorage` object that mimics filesystem semantics
* Adding a `ConfigStorageInterface::initializeStorage()` call to let the storage backend create directories or database tables, as appropriate
* Cleaning up some of the install process to be more storage agnostic
* Reorder the bootstrap sequence to put `BOOTSTRAP_DATABASE` on top, before `BOOTSTRAP_CONFIGURATION`.

This PR is based on and includes the patch for backdrop/backdrop-issues#2275.

The tests are, of course, a complete mess.

**Comment by sentaidigital** [comment]
Test are fixed by adding a new Bootstrap stage: `BACKDROP_BOOTSTRAP_DATABASE_EARLY`, which has all the database bootstrap code, minus the test prefix validation.  The validation remains in the original `_backdrop_bootstrap_database()` function.

Still some work to be done (import/export is broken, for starters), but I'd like someone else to look at the extra bootstrap stage.

**Comment by quicksketch** [comment]
Thanks @sentaidigital for making this issue! It's an item I've been thinking about on and off for quite a while. Using a database storage could make a lot of sense in various hosting situations. You're totally right that in a high-availability scenario, the database is much more accessible than the file system (though that should be shared too, and the options are a lot better now than they were a few years ago, e.g. Gluster and Amazon EFS).

Pantheon in particular could benefit from this change, because it would make it so that if you downsynced your database, it would also include your configuration. The configuration directory isn't easily accessible on Pantheon anyway, so storing the configuration in the database might give you a multiple wins.

The downsides are that database-storage is less accessible and visible than the file system for developers. Config files being easily accessible makes the entire operation of Backdrop seem more transparent and simple than it all being stored in the database.

But overall, I think this is a great option to have, just as long as the default is the file system with the option to switch to the database. Speaking of which, do you have any ideas on how you could change your config storage type other than reinstalling? Seems like it might be a good candidate for a Drush issue.

> I'd like someone else to look at the extra bootstrap stage.

I left a full review on the PR. The extra bootstrap phase in particular I think we could improve. Here's my comment from the PR:

> Rather than adding a new bootstrap phase, would it be possible to just make the config bootstrap phase "upgrade" the boostrap level in the contructor for the new database config storage class? This is the same thing that BackdropDatabaseCache::__contstruct() does. Even though the cache backend is needed at boostrap level 2, it upgraded the level to 3 so that it can serve the page cache from the database.

For illustration, here's the code from `BackdropDatabaseCache()`:

```php
  /**
   * Constructs a new BackdropDatabaseCache object.
   */
  function __construct($bin) {
    // All cache tables should be prefixed with 'cache_', except for the
    // default 'cache' bin.
    if ($bin != 'cache') {
      $bin = 'cache_' . $bin;
    }
    $this->bin = $bin;

    // Bootstrap the database if it is not yet available.
    if (!function_exists('db_query') || backdrop_get_bootstrap_phase() < BACKDROP_BOOTSTRAP_DATABASE) {
      backdrop_bootstrap(BACKDROP_BOOTSTRAP_DATABASE, FALSE);
    }
  }
```

Seems like we could use that same bootstrap upgrading code in `ConfigDatabaseStorage::__construct()` and it would do the job of making the database accessible earlier if you're using database-stored configuration.

**Comment by sentaidigital** [comment]
If the bootstrap upgrade works, I have no objections.

> But overall, I think this is a great option to have, just as long as the default is the file system with the option to switch to the database. Speaking of which, do you have any ideas on how you could change your config storage type other than reinstalling? Seems like it might be a good candidate for a Drush issue.

It's just JSON in the database record.  It should be trivial to export from file and import into the database.  Even if we add a way to do it in core, a Drush command would still be a good solution.  Either way, migrating from one to the other is beyond the scope of this issue.  

I'd like to make the installer UX updates a separate issue, too.  Right now, you have to specify the database in settings.php before you start the install, which is good enough to let more advanced site builders play with and break it while the UI is built.

Regarding the JSON part (see comments in PR):

Honestly, I'm not happy about the encoded settings part of this patch.  My original plan was to store one setting per row in the database (more like the variables table) so developers and sites admins could change individual settings easily with simple SQL statements.  I've had to do this a number of times to fix something broken in a D6 or D7 site.  (In fact, I submitted a [patch for Domain Access](https://www.drupal.org/node/1737144) to do so.)  Editing a scalar or simple array that's been serialized is doable.  Editing a large array that has been serialized is error prone, to say the least.

However, the data is sent to the `ConfigStorageInterface` is everything in the file, not just the deltas, which means `write()` would have to either delete the entire block of settings before writing out the new data or do some clever merging to prevent unnecessary writes to the database.

Maybe we can change the semantics from full file to individual setting in 2.x, and leave the JSON functions duplicated until then.

**Comment by sentaidigital** [comment]
This does not work: 

```php
    // Bootstrap the database if it is not yet available.
    if (!function_exists('db_query') || backdrop_get_bootstrap_phase() < BACKDROP_BOOTSTRAP_DATABASE) {
      backdrop_bootstrap(BACKDROP_BOOTSTRAP_DATABASE, FALSE);
    }
  }
```

But this does:
```php
    // Bootstrap the database if it is not yet available.
    if (!function_exists('db_query') || backdrop_get_bootstrap_phase() < BACKDROP_BOOTSTRAP_DATABASE) {
      require_once BACKDROP_ROOT . '/core/includes/database/database.inc';
    }
  }
```

The call to `backdrop_bootstrap(BACKDROP_BOOTSTRAP_DATABASE, FALSE)` does not work because it never calls the `_backdrop_bootstrap_database()` function.  When the constructor calls `backdrop_bootstrap()`, bootstrap does not update the `$final_phase` (bootstrap.inc:2668) because `$new_phase` is `FALSE` and the current `$stored_phase` and `$final_phase` are both 0 (`BACKDROP_BOOTSTRAP_CONFIGURATION`), so the while loop (bootstrap.inc:2673) never runs.

The bootstrap upgrade _does_ work if we change that while loop from:
```php
    while ($phases && $phase > $stored_phase && $final_phase > $stored_phase) { ... }
```
to 
```php
    while ($phases && ($phase > $stored_phase || $final_phase > $stored_phase)) { ... }
```

**Comment by mikemccaffrey** [comment]
> The downsides are that database-storage is less accessible and visible than the file system for developers. Config files being easily accessible makes the entire operation of Backdrop seem more transparent and simple than it all being stored in the database.

I was initially skeptical of D8's decision to move the active configuration back into the database, but after using it for a while, I'm starting to think it was a good idea. We really don't want to encourage manual editing the active configuration of a site directly. We should be telling people to export their configuration, edit the files, and then import the changes.

> It's just JSON in the database record. It should be trivial to export from file and import into the database. 

Should we be storing JSON in the database? Even if we are importing and exporting JSON to files, it seems like it might make more sense to store serialized PHP objects in the database records. Again, we don't want people making changes to the active configuration of the site, and should instead be leading them towards the export/change/import workflow.

**Comment by sentaidigital** [comment]
The most recent `hook_update_N()` from #2222 revealed some issues in the update code.  A couple patches addressing those issues are now here.  

The `exportArchive()` and new `importArchive()` member functions are updated, but not tested.

**Comment by sentaidigital** [comment]
@quicksketch if you have no objections, I'd like to move some of the changes here to the #2275, including any additions to `ConfigStorageInterface` and their implementation in `ConfigFileStorage`.

**Comment by sentaidigital** [comment]
Removed the API additions that are now in #2275 and cleaned up the patch to remove the reverted bootstrap changes.  This should make the patch easier to review and do a better job keeping both issues in scope.

**Comment by jenlampton** [comment]
Hi @sentaidigital and thanks for this PR.  Unfortunately, I'm not yet sure this is the kind of change we should be making in Backdrop. My instincts are telling me that this kind of feature addition goes directly against Backdrop's principles.

> Makes it easier to synchronize config changes among multiple web servers in a load balanced environment.

This is an edge-case scenario. Changes specifically for this group should not be put in Backdrop core without careful consideration. If we start adding features that benefit the only minority of sites, we are going to end up with code bloat and increased complexity that adversely affects the majority. 

> Makes it easier to capture the entire state of an environment for debugging elsewhere.

You need to capture both the files and the database to get an exact snapshot. If you keep the config in your files directory (as it is by default) there's nothing else to download. Only when you move the configuration out of the files directory do you now need to capture a third thing. And even so, is it really that much more work to grab a config directory than it is to grab the files directory?

> Pantheon in particular could benefit from this change, because it would make it so that if you downsynced your database, it would also include your configuration.

But this isn't always wanted. How would you then down-sync your content only (without configuration)? I don't think Pantheon would add a checkbox for "Database with config" and one for "Database without config" in their UI. We separated content from configuration so that these things could be managed separately, and this feature would defeat that purpose.

I may be wrong about the possible benefits to our majority use-case, but I'd be curious to hear more perspectives. 

Either way, I don't think this is a decision we should rush into, and I'm not sure why this issue was tagged with 1.6. We need to more carefully consider the implications of increasing the complexity and decreasing the visibility in exchange for a feature that may only benefit the 1%.

**Comment by sentaidigital** [comment]
This PR doesn't mandate how developers or vendors should approach configuration management in Backdrop.  It only provides an option to store configuration in a database instead of in flat files.  Developers, regardless of their savvy, can still use the default `ConfigFileStorage` to store configuration in files.

Nor does this require that database configuration be stored in the main database.  Defining a second database connection `$databases['config'] = array(...)` and setting the configuration path to `db:/config/active_config` would preserve the separation of content and configuration.

Nor does it require the staged configuration to live in the database.  Developers can still edit the configuration files in the staging directory and then import it in to the active config in the database.  Actually, this is how I expect a site using database configuration to be managed.

The implementation here is completely backwards compatible.  Existing installations will continue to run with their file-based configuration, exactly as before.  In fact, until a future PR updates the install process, the option to use the database for configuration will require someone to set it in their `settings.php` before running the install process.

There's not a lot of complexity, either. The PR amounts to one additional class and a switch to choose between `ConfigFileStorage` and `ConfigDatabaseStorage`.  It's a textbook example of how OO development should be done.  The only other piece is a small patch in `bootstrap.inc` to ensure the database is initialized in time to load the config from the database.  If the PR looks like more, it's because the first nine commits are the commits from #2275.  Splitting the PR into two, one to refactor and clean up the existing, one to add a new storage class, was done to make them both easier to review.

Yes, the load balanced environment argument is an edge case, and isn't the best argument to sell this.  Security is a better one.  Databases tend to be behind firewalls and only accessible by the webheads, which protects the configuration better than an easily misconfigured web server directive protecting one directory in an otherwise published `files` directory.

There is also a DX argument to be made.  I generally don't need to copy down the files directory, because I can look past the missing images or broken download links.  Grabbing the entire configuration and data in a single operation is easier.  Which is not to say grabbing both files and database is hard, only that one operation is easier than two operations, and harder to mess up.

**Comment by jenlampton** [comment]
I'm still concerned about adding a feature that only benefits some extreme minority of sites, increases the amount of code we'd need to maintain, increases the overall complexity of the system, and decreases the visibility of how configuration files work when the feature is used. All of those things go against Backdrop's philosophy.

That concern doesn't mean we shouldn't do this, or that it's a bad idea. All it means is that this decision needs to be considered carefully and not just merged in because the code looks good.

**Comment by serundeputy** [comment]
I don't think we/Backdrop should relegate high performance websites as an edge case.  It seems perfectly reasonable to me that there will be sites that need load balancing and high performance techniques and it seems likely to me that those will be the ones that give Backdrop CMS the confidence boost for others to adopt.

The tipping point for me is performance.  @sentaidigital can we get some metrics: 

* single install sites with config in the DB vs. config in files
* load balanced config in DB vs. config in files

If we can demonstrate that the performance is better or on par with the current situation then I think this is a good add with no impact for people that spin up Backdrop on shared hosting.  They will never know about this setting until/and if they ever need it.

I know we target small to medium size businesses and non-profits.  I've worked on several non profit sites that are high traffic with load balancing and cacheing in place. 

>Pantheon in particular could benefit from this change, because it would make it so that if you downsynced your database, it would also include your configuration. The configuration directory isn't easily accessible on Pantheon anyway, so storing the configuration in the database might give you a multiple wins.

* we can down sync the files dir to on Pantheon.  It is an option on the dash for down syncing DB and a separate option to down sync files.  So that should work fine in our current scenario.

That is my 2¢; let's see if we can get those metrics.

**Comment by jenlampton** [comment]
> I don't think we/Backdrop should relegate high performance websites as an edge case.

Why is less than 1% of sites **not** an edgecase? That's the exact definition of an edgecase! Yes, maybe high-profile sites using backdrop will sell others on using it, but that doesn't mean their requested changes need to go into core. (not that this is one, but you get the point)

I don't care so much about performance in this case because it's a setting that can be enabled / disabled, though seeing the metrics might be interesting. There should be a cost associated with keeping the config in the db, but we've never been able to measure it (before now). 

FTR: I'm less concerned with this particular issue and more concerned with how Backdrop will handle these edge-case feature requests, and how we do the cost/benefit analysis to figure out if/when requests like this should go in.

**Comment by Gormartsen** [comment]
> - Makes it easier to synchronize config changes among multiple web servers in a load balanced environment.
- Makes it easier to capture the entire state of an environment for debugging elsewhere.

Hi @sentaidigital

I am doing a lot of CI/CD and multiple nodes systems.
I found that keeping configs in json files actually do make deployment and testing much easier.

I keep configs in github repository as a part of the project and all  config changes on production brunch get deployed to multiple servers at the same time.

Good part of it is that you can rollback configuration without affecting database data.

Also you can track who and when did changes to specific config and what was a reason.

I suggest you to explore this option.
When properly adopt, it makes development for big projects more secure and stable.
And much easier.

**Comment by Gormartsen** [comment]
@jenlampton - I usually works on edge-case and performance critical tasks.
I will be more than happy to do measurements and tests for this type requests and if necessary provide alternative ideas to handle it.

**Comment by sentaidigital** [comment]
Quick datapoints:

Test run of this PR using PHP 7 and file-based config: [Link](https://zen.ci/backdrop/backdrop/test/test-php70-backdrop_backdrop_1613-29cf745) run_tests.php takes 4m05s
Test run of this PR using PHP 7 and db-based config: [Link](https://zen.ci/backdrop/backdrop/test/test-php70-backdrop_backdrop_1613-1b77db4) run_tests.php takes 3m42s with 1 exception
Same with fixed exception: [Link](https://zen.ci/backdrop/backdrop/test/test-php70-backdrop_backdrop_1613-e873268) run_tests.php takes 3m43s

There are a lot of variables here.  The server may have been under higher load during the stock tests, which were run several days ago.  @Gormartsen, I'd appreciate your help coming up with more reliable numbers.

**Comment by Gormartsen** [comment]
@sentaidigital
Tnx for numbers!
I will be able to do extra speed tests only after December 7. (I am still in Peru, doing traveling)

I saw latest commit. We do need to came up with proper strategy to do all tests under files and db configs to avoid any possible exceptions in future.

**Comment by Gormartsen** [comment]
Ps: each test run on new google cloud instance, so here is no effect on tests speed if you run many tests simultaneously

**Comment by sentaidigital** [comment]
The patch currently uses `array_column()`, which doesn't exist before PHP 5.5.  There is a compatibility function [here](https://github.com/ramsey/array_column/blob/master/src/array_column.php) under the MIT license.

What is core's policy on including such files with non-GPL'd code?

**Comment by Gormartsen** [comment]
@sentaidigital I created PR backdrop/backdrop#1661

I added small patch to move config/active to /dev/shm (memory based partition).

This way we are comparing speed for configs stored in database in memory (MySQL store database in /dev/shm/mysql on text boxes ) and in files in memory (/dev/shm)

Config in MySQL:
php7 -  3m 43s [link](https://zen.ci/backdrop/backdrop/test/test-php70-backdrop_backdrop_1613-e873268)
php5 failed due to `array_column()`

Config in files in memory:
php7 - 3m 29s [link](https://zen.ci/backdrop/backdrop/test/test-php70-backdrop_backdrop_1661-d36a8ce)
php53 - 7m [link](https://zen.ci/backdrop/backdrop/test/test-php53-backdrop_backdrop_1661-d36a8ce)

Config in files on SSD drive (backdrop/backdrop#1656):
php7 - 3m 42s [link](https://zen.ci/backdrop/backdrop/test/test-php70-backdrop_backdrop_1656-af3dab9) 3m 39s [link](https://zen.ci/backdrop/backdrop/test/test-php70-backdrop_backdrop_1612-6603354)
php53 - 7m 26s [link](https://zen.ci/backdrop/backdrop/test/test-php53-backdrop_backdrop_1656-af3dab9)


Based on this results I can say that there is no speed improvement from Database based configs for tests.

Speed stay +- the same. 
Even using /dev/shm for files is not affecting performance too much.

Numbers could be different on SATA based servers due to bigger seek time (But FS usually cache most requested files as well)

**Comment by jenlampton** [comment]
Is there any reason this can't be done in contrib?

**Comment by Gormartsen** [comment]
I like an idea of adding API changes to config (like we have for cache) to make it possible to replace default ConfigFileStorage.

For example, XCacheConfigStorage or even MemcacheConfigStorage could be a good contrib modules.

**Comment by jenlampton** [comment]
I do too. Then this feature (or anything esimular needed in the future)
could exist in contrb, and we'd be adding less that isn't for our key users.


On Dec 4, 2016 5:40 PM, "Gor Martsen" <notifications@github.com> wrote:

I like an idea of adding API changes to config (like we have for cache) to
make it possible to replace default ConfigFileStorage.

For example, XCacheConfigStorage or even MemcacheConfigStorage could be a
good contrib modules.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/backdrop/backdrop-issues/issues/2277#issuecomment-264749695>,
or mute the thread

**Comment by sentaidigital** [comment]
> Is there any reason this can't be done in contrib?

Bootstrap.  We have to be able to load a module to load the configuration that we don't know is enabled until we load the configuration.

**Comment by Gormartsen** [comment]
@sentaidigital - same as cache. So yes, we do need patch core to make it possible to replace default config class via settings.php for example.

**Comment by Gormartsen** [comment]
upd: I mean, we need to change core to make possible something like this:

settings.php
```
$settings['config_default_class'] = 'ConfigDatabaseStorage';
$settings['config_backends'] = array('modules/config_db/config_db.class.php');
```

**Comment by sentaidigital** [comment]
Adding `$settings['config_backends']` is a big enough change to warrant its own issue.

**Comment by Gormartsen** [comment]
@sentaidigital basically I am proposing to change core in a way to make it possible to use contrib module for ConfigDatabaseStorage

This way we minimize core changes and you can have contrib module for your needs.
Kinda compromise

**Comment by sentaidigital** [comment]
A contrib module will eventually need a way to tell `install.php` that `ConfigDatabaseStorage` is available as a backend, before any modules are initialized, even core modules like system or user.  I know that changes to support install time selection are not in this patch, but this patch also doesn't make such changes harder.

It's not that this is impossible, but as I mentioned in the [review of the PR](https://github.com/backdrop/backdrop/pull/1613/files/bd6cccf4b420cda6052af778420da8f70c7e21cb#diff-71ce5739c2a63620ff2ab683eceae1aa), "doing so would radically change the scope of this issue."

Even a "simple" patch to add `$settings['config_backends']` has ripple effects on parts of bootstrap and `config_get_config_storage()`, which needs to know the difference between `db://active_config` vs `files/active_config`.  Setting `$settings['config_default_class']` doesn't work because the active configuration may be in the database while the staging configuration is in files.

We can use the `$settings['config_backends'][$classname] = "path/to/class.inc"` to specific the file and the class name, add public static member functions that return `TRUE` if they can parse/validate the config path, and another to tell `install.php` whatever it needs.  (I haven't even looked at that code.)  Of course, the `$settings` stuff all needs to be set before running `install.php`, which is an awful lot to ask of less savvy users.

I'd prefer a solution where a site builder needs only drop in the module to get the new storage class, and use the same solution to support new `DatabaseConnection` classes and `BackdropCacheInterface` classes, too.  The `settings.php` solution, while it avoids certain chicken-and-egg issues, always felt clunky to me and required an advanced understanding of the system to even know the `*_backends` exist.

But, it is the chicken-and-egg issues that make it so hard.  We can't use the autoloader because it's not even registered until the end of `BOOTSTRAP_DATABASE`.  How early can we scan every module's .info file and how bad of performance hit would that cause?  Maybe `install.php` should do the scan and symlink the backends into a particular directory that early bootstrap scans or write out a `backends.php` next to settings.php, but then simply adding a module doesn't add the backends, even though they are clearly in the .info file, which is not how anything else in Backdrop works.  Do we have to set a special directory and class of module for backends like we do themes and layouts?

The core of the problem is backends are different from other contrib modules.  Config, Database, and Cache classes available in contrib introduces a device driver-like class of contrib module.  It's been clunked in with Cache modules so far.  This has been less of an issue, because the sites that need advanced caching have more savvy developers building and maintaining them, but it's always been a one-off.  It's more of an issue for sites that would like to use other simple databases, such as SQLite or the desktop version of SQL Server (SQL CE?), nor for site builders used to the D6, D7, and current D8 practice of storing configuration in the database.

These are issues I was hoping to defer to a later time, because they're big enough to warrant their own issue.  Compared to a low-level plug-in system, adding a second class implementing `ConfigStorageInterface` plus a bit of plumbing to support it is kinderspiel.

**Comment by jenlampton** [comment]
> What is core's policy on including such files with non-GPL'd code?

@sentaidigital MIT is GPL-compatible, so adding that function would be fine.

> A contrib module will eventually need a way to tell install.php that ConfigDatabaseStorage is available as a backend, before any modules are initialized, even core modules like system or user.

Could we still use the file system for config during install, and simply move all the config into the DB when installation is complete?

> The core of the problem is backends are different from other contrib modules. Config, Database, and Cache classes available in contrib introduces a device driver-like class of contrib module. It's been clunked in with Cache modules so far. This has been less of an issue, because the sites that need advanced caching have more savvy developers building and maintaining them, but it's always been a one-off.

The same logic works for this feature too, the sites that need db config storage also have more savvy developers building and maintaining them.

> It's more of an issue for sites that would like to use other simple databases, such as SQLite or the desktop version of SQL Server (SQL CE?).

We've dropped support for anything other than MySQL (and MariaDB) from Backdrop, because that is also an extreme edgecase. Sites that would like to use these other simple databases will not be using Backdrop CMS.

> These are issues I was hoping to defer to a later time, because they're big enough to warrant their own issue. Compared to a low-level plug-in system, adding a second class implementing ConfigStorageInterface plus a bit of plumbing to support it is kinderspiel.

Even though the code in this PR may be minimal, we don't usually add features to Backdrop unless we're sure that at least 80% of sites will benefit. Here, that is certainly not the case. However, sometimes these decisions are not always so black and white. There are other factors that may influence the decision, and one thing we should consider is what would need to be done in core to make this possible in contrib. 

If comparing the two approaches reveals that what has to be done to core to support this kind of feature in contrib is far more complex and has a worse effect on performance than adding it to core, that may encourage us to make an exception in this case.

**Comment by klonos** [comment]
> If comparing the two approaches reveals that what has to be done to core to support this kind of feature in contrib is far more complex and has a worse effect on performance than adding it to core, that may encourage us to make an exception in this case.

This sounds reasonable to me for non-80% cases.

**Comment by mikemccaffrey** [comment]
> Security is a better one. Databases tend to be behind firewalls and only accessible by the webheads, which protects the configuration better than an easily misconfigured web server directive protecting one directory in an otherwise published files directory.

This is a very good point. Our current system of adding the long hash in the config file directory is kinda a kludgy workaround, and it is never good to rely on security through obscurity.

> There is also a DX argument to be made. I generally don't need to copy down the files directory, because I can look past the missing images or broken download links. Grabbing the entire configuration and data in a single operation is easier. Which is not to say grabbing both files and database is hard, only that one operation is easier than two operations, and harder to mess up.

This is my experience with developing drupal sites as well. A single database sync to get the latest content and configuration is much easier than doing a file sync as well, especially on sites that have a large number of files. In fact, I most often use stage_file_proxy and don't sync the files directory to dev.

> Even though the code in this PR may be minimal, we don't usually add features to Backdrop unless we're sure that at least 80% of sites will benefit. Here, that is certainly not the case. However, sometimes these decisions are not always so black and white. There are other factors that may influence the decision, and one thing we should consider is what would need to be done in core to make this possible in contrib.

I would not classify this issue as adding a feature to core. I'm leaning towards thinking that this should be our default config storage solution, especially since it would bring us in closer parity to how config storage works in D8, so users don't need to learn yet another way to manage things. 

I'd like to propose that we add this storage option to core in the next release or two so we can try it out, make it the default config storage option if it works well in another release, and then remove file storage as an option in 2.x unless there is a clear reason to continue offering that as an option.

**Comment by Gormartsen** [comment]
@mikemccaffrey - I have strong feeling that BD has to go his own path. 
I see that getting anything that belong to structure to files is essential to backdrop future.

Using json files to store state and structure data is my favourite feature for next reason:

- I can store settings (configs) in github. 
- Easy to track changes and revert back
- easy to deploy settings (configs) on any environment (dev, stage, production) without affecting data.


I don't use UX to import export settings (configs) at all.

Quick  example:
Task: I need to update layout settings.
My steps:
1 - update it on my local env
2 - git commit layout config changes to github
3 - deploy config to production (via zen.ci, circleci or any other CD system or even just `git pull`)
Done.

**Comment by jenlampton** [comment]
I agree with @Gormartsen. There are other solutions to the issues mentioned above that don't have the downsides of this approach. There won't be very much overlap with Drupal 8 in the long run, and we need to make our own decisions. The people we are targeting with Backdrop are the people who would benefit the least (and suffer the most) from this kind of change. We need to be careful when making this decision that we aren't doing what we want personally, but are instead acting in the best interests of the people this project is intended for. 

> Backdrop values site builders over coders. Backdrop values contrib developers over core developers.

There are tools for the experts that can be used for config syncs and db syncs and to make the act of copying files easier. Instead, we should be focused on how easy it is to get a non-developer into using Backdrop, and not be throwing extra hurdles in their way to benefit the pros.

**Comment by sentaidigital** [comment]
> Instead, we should be focused on how easy it is to get a non-developer into using Backdrop, to understand how it works, and not be throwing extra hurdles in their way to benefit the pros.

If we can focus on the patch, what extra hurdle is being added here?  The patch does _not_ change the default configuration scheme, nor does it change the install process.  Anyone who isn't reading the release notes won't even know they _can_ store configuration in the database.  

A future patch will add radio buttons to the installer that read "Config in files" and "Config in database", with the former selected by default.  We can bury it in a collapsed "Advanced options" fieldset or vertical tab.  The current case will be preserved, another option will be available.  A single radio element is not a hurdle.

Ensuring the common case is easy is a great goal.  Denying pros the tools they want and need because we're afraid the tools may be confusing turns away the very people we want helping and mentoring the non-developers and the rookie developers.  In the end, it just hurts the community.

There should be room in the Backdrop philosophy for the classic Perlism, "easy things should be easy, difficult things should be possible."

**Comment by Gormartsen** [comment]
I feel sad that I miss developer meeting. That was a hot and important one.
Thank you for your opinions guys, it helped me to shape mine.

My opinion can be counted as a DevOps, a hosting company and advanced developer opinion.

- About not tech users.
  - Backdrop CMS (Drupal 6,7,8 and etc) is black-box to them. Doing import and export of database or configs is something that they are not going to use by themselves. 
  - UI&UX is what important to them. It should be easy to find and change anything via web Interface. As long as we can success on this field, they are happy about BackDrop CMS.
  - they do not migrate a site from one to another hosting by doing backups. They ask tech support or developer to do so.

- dev Beginners
  - the more code is readable the easier beginners can grow. From this side having configs in a specific table as serialized object is more complicated than having a single file that is human readable.
  - no matter how hard we try to prevent users from changing config.json file directly - it is going to happen just because it is much **simpler and faster**! You just need to know json syntax. 

- DevOps
  - devOps is getting more popular and less complicated. Any CMS that is going to be devOps friendly will be more successful. Just because it helps to be more competitive. Configs in files are more devOps friendly. Configs in Database are not.
  - it is very important to deploy new feature without affecting user data. Physical separation user data from structure data is very important.
  - automated updates on a database are much complicated  than updates on a specific file. Files can be a part of any CVS and configuration conflicts are more transparent.

- Hosting (I am doing Drupal hosting since may 2008)
  - MySQL database is an expensive resource. We do recommend filecache module to our users to speed up performance and save money on hosting expenses. Basically, the less site call database the faster it is. MySQL (Percona) is still not friendly to SSD and use memory when it is not necessary. SSD is common now and provide incredible IO performance. You can test it by installing [backdrop filecache](https://github.com/backdrop-contrib/filecache) and see results. 

Going back to feature request. I like work that is happening in #2383. This feature is contrib developers friendly. It is matching our philosophy.

**Comment by jenlampton** [comment]
> If we can focus on the patch, what extra hurdle is being added here?

None, that was addressing the "make it default" comment from above.

> There should be room in the Backdrop philosophy for the classic Perlism, "easy things should be easy, difficult things should be possible."

I agree with this 100%, that's why I asked about what's possible to do in contrib. If it's possible (even if it's a hack) that decreases the need to get this change into core, at least any time soon. I'd love to see a contrib module first to gauge interest, then a core issue if it turns out to be in high demand.

**Comment by quicksketch** [comment]
The issue for more consistent accessing of the file-based config storage has been merged at https://github.com/backdrop/backdrop-issues/issues/2275.

It might be worth it to reroll https://github.com/backdrop/backdrop/pull/1613 to get an idea of how much effort is now involved here. With the API additions all separated out, it may make the scope of change here more clear.

Thanks again @jlfranklin for your iterative approach in separating these tasks out for evaluation one at a time. I don't think this will be able to make it for 1.6.0, but we're going to be in a much better place either for 1.7 or your new Silkscreen performance fork.

**Comment by jlfranklin** [comment]
I don't mind refiling the PR, but I thought this one was already declined for core.

**Comment by jenlampton** [comment]
> I'd love to see a contrib module first to gauge interest, then a core issue if it turns out to be in high demand.

It was declined for core **right now**, but knowing how much change is required will help us re-evaluate in the future. If this is something some people want, then we'll need to be able to measure the change required so we can decide if it's worth the cost for everyone else.

**Comment by alexfinnarn** [comment]
Hello folks! I've tried out this PR and like the simplicity of only having to change one line in settings.php and the code changes seem nice and tidy. 

I'll go through the steps I took to test:
- Go to the Backdrop CMS homepage and follow the link to create a site on Pantheon.
- Clone down the code, apply the PR, and make the change for active config directory in settings.php.
- Push up the changes and install the site on Pantheon in the dev environment.
- Initialize the test and live environments.
- Make a text block content type and add a layout at "/contact" including one "text block".
- Clone back the database from live to dev.
- See the layout, content type, and "/contact" page look the same as the live site.

I also repeated the same process with the normal `ConfigFileStorage` way of working with config. My main reason for testing out this PR was running into this issue on Pantheon: https://github.com/backdrop-ops/backdrop-pantheon/issues/27. The issue seems to be intermittent, and I can't point to any reason why other than something with the Terminus plugin for cloning code or how Valhalla works for projects that aren't Drupal or WP. If someone is evaluating Backdrop on Pantheon and runs into that issue, though, it might turn them off of their investigation. 

Where I work, we run a service for hundreds of sites (some would say 1000+) on D7 and are used to updating config using update hooks and Features when deploying code. Due to the nature of our current infrastructure, it would be far easier for us to store the config in the db and remove a blocker. We don't have a process of working with config in files and that change would be difficult for us to make. Granted, doing things this way goes against moving config into files, and I can't really find anything in https://backdropcms.org/philosophy pointing to changes like this one. 

So, what is the feeling on this issue after almost two years have gone by? I've read the comments about pros and cons to this change but it only looks like adding a couple of conditional statements and a class that follows a well-defined interface. I can't comment too much on the code or performance considerations, though, only that the PR worked for me in limited testing and I like the idea. Are there specific types of testing I should be doing?

One interesting question to point out on Backdrop philosophy page:

> 3. Focus: Include features for the majority. Backdrop core should only include features and tools that benefit the majority of sites that are running it.

What if a few groups who run services with hundreds of sites migrate to Backdrop and they now make up the majority of sites. Does this change then become more in-line with project philosophy? Even going back to the project focusing on site builders...what if the site builders work for these companies? What if the editors prefer to work with db storage and they all work for these companies? Maybe not a topic for this issue in particular, but it is a question that we think about while evaluating Backdrop coming from Drupal and the promise of the PMC being better than BDFL governance. If larger groups go all-in on Backdrop and are still "not the users Backdrop is intended for", then they're probably not going to want to join in and potentially ruin the community's feel.

**Comment by jlfranklin** [comment]
@alexfinnarn , have you considered running [Silkscreen CMS](https://github.com/silkscreencms/silkscreen/releases), the fork of Backdrop that includes those patches and continues to track the other changes?  It is intended to be a drop-in replacement for Backdrop, like Pressflow was for Drupal 6.

**Comment by alexfinnarn** [comment]
@jlfranklin I know we've looked at Silkscreen but have been more focused on evaluating the main Backdrop project against other alternatives for a migration path off of D7.

**Comment by klonos** [comment]
@alexfinnarn without #285 in place, we cannot be 100% sure of what the majority uses and how.I think that so far we have been going by gut feeling so to say. There have been some features added in core for which we have absolutely no hard *proof* that they would serve the majority or not.

It is true that the project started with only a handful of people, and it was easy to discuss and see what "the majority" felt like. As we grow, I think that #285 will become a necessity if we are to get a good feel of what works for most people. Having said that, there was also no hard proof that everybody hated the overlay.module 😜, but we did remove it. My point is that I feel that we have been doing a good job so far.

Speaking of #285, even when I was reading https://www.colorado.edu/webcentral/2018/04/11/change-my-view-d8-isnt-best-upgrade-path-1000-d7-edu-sites I was thinking "Well something like that would surely throw our stats off" 😄 ...I am sure that cases like that would come along over time, but I have hope that the community will grow enough that 1000 sites would only be a minority (🙏 😄)

**Comment by jenlampton** [comment]
The PMC has voted to make config storage swappable. 

It's unclear to me whether the ConfigDatabaseStorage class should be included in core or needs to be provided by contrib, but we have agreed that needs to be possible to choose database storage. 

I will ask for clarification in the PMC issue.

**Comment by klonos** [comment]
I realize that there hasn't been any movement here recently, but I don't want to dismiss this by jumping to conclusions re demand for this feature. I think that at the very least core should make this easy/possible in contrib.

With #285 now done, and https://backdropcms.org/project/backdrop/telemetry providing us a decent amount of data (I'm sure we'd love it if the sites reporting back metrics were more - that will come with time), wondering if it'd be worth it capturing usage stats for Silkscreen 🤔 ...it would certainly help us make better-informed decisions in matters like this one here I think.

**Comment by klonos** [comment]
Hey @jlfranklin are you still actively maintaining the project by the way? ...asking because I see that there's a broken image in the homepage of https://www.silkscreencms.org and https://github.com/silkscreencms/silkscreen is reporting v1.21.5 as the most recent.

Also, do you have any rough idea of usage of the project? (have you changed anything related to where Silkscreen installations send their stats/telemetry, or is it still the main b.org site?)

**Comment by kiamlaluno** [comment]
Saving the configuration data in the database would help those sites with a low disk usage for files, if there are. In my case, the database disk usage is limited to 50 GB, which is the same limit I have for files. I cannot exclude there are planes with a lower disk usage for files, which give more space for database tables.

**Comment by klonos** [comment]
Although a valid argument @kiamlaluno, still, how much space does config use up? A typical Backdrop site on my local has something like ~500KB of config files. So the disk space savings are kinda negligible.

**Comment by laryn** [comment]
I'm just coming to this issue but interested if it might help on Pantheon sites in particular where it seems to be the consensus that loading config from the writable `/files` directory is dreadfully slow. Perhaps this would make things quicker?

**Comment by herbdool** [comment]
@laryn see the comment above https://github.com/backdrop/backdrop-issues/issues/2277#issuecomment-437419494

**Comment by laryn** [comment]
I am starting some early testing of this patch. I have a sandbox site on Pantheon that I wanted to use rather than starting from a fresh install, so I did the following:

1. Make sure the sandbox site is in Git mode (not SFTP)
2. Apply the patch
3. Run the following code (I just used the `devel` module's `Execute PHP code` and did it right in browser):
```
$config_db = new ConfigDatabaseStorage('db:/default/active_config');
$config_db->initializeStorage();

$config_file_groups = config_get_prefix_groups();
$group_configs_loaded = [];
foreach ($config_file_groups as $config_file_group) {
  $group_configs_loaded += config_load_multiple(array_keys($config_file_group));
}
foreach ($group_configs_loaded as $config_name => $config_loaded) {
  $config_db->write($config_name, $config_loaded);
}
```
4. Update `settings.php` to change the active config location:
`$config_directories['active'] = 'db:/default/active_config';`
5. Flush caches and try to change some config settings (still in Git mode, not SFTP mode): success!

**Comment by quicksketch** [comment]
Can we try to come up with better syntax for specifying a DB-backend? I don't really like repurposing `$config_directories`, since they're really not directories at this point. I left feedback some time ago that was similar: https://github.com/backdrop/backdrop/pull/1613/files#r83577378

But I don't think it's that big a scope expansion and since we won't be able to easily change this syntax later, we should try to have it be as accurate/consistent with terminology as possible.

What I would expect to see based on our current syntaxes for cache backends is that we add specifiers like this:

```
$settings['config_active'] = 
```

**Comment by herbdool** [comment]
I agree about non-default database being a follow-up. Also I don't think we even need a setting for the staging class, do we? I don't think anyone so far has demonstrated a need for storing that in the database too.

**Comment by laryn** [comment]
With thanks (and apologies) to @jlfranklin / @sentaidigital I've filed a new PR on top of the earlier one:

https://github.com/backdrop/backdrop/pull/4453

@herbdool I didn't see your comment until now. See what you think of the modifications in the latest PR, which does technically allow that option yet.

**Comment by herbdool** [comment]
@laryn I had left the comment right after your commit. Let's leave it in for now until there's more feedback. I've also left a review. Some minor fixes and one major: I don't think `ConfigDatabaseStorage::importArchive()` works.

**Comment by laryn** [comment]
@herbdool I think this is ready for another round of testing.

**Comment by laryn** [comment]
Just got this when trying to clone down a Pantheon site locally but had not updated the database yet (should be adapted with database-specific language in this case):
<img width="889" alt="image" src="https://github.com/backdrop/backdrop-issues/assets/871421/9f359e57-ca19-4494-9730-b2e2f0395c53">

**Comment by laryn** [comment]
Ready (again).

**Comment by herbdool** [comment]
@laryn the error message looks better, but I wonder if it should still mention `$config_directories` since that could still be a source of problem for the default configuration. Makes the message more convoluted but I don't see a way to avoid that.

**Comment by herbdool** [comment]
I was able to successfully export/import individual configs and also export/import a whole tarball.

**Comment by herbdool** [comment]
When start up the site in lando I see this error:

> Exception: The active configuration location is either empty or missing crucial information. Check that the $config_active_class variable is correct in settings.php. in _backdrop_bootstrap_configuration() (line 3051 of /app/core/includes/bootstrap.inc).

The site otherwise works fine. First, the variable should be `$settings['config_active_class']` and second, I get the error even if it is set properly. Perhaps at that point in the bootstrap the database is not yet available?

**Comment by kiamlaluno** [comment]
`_backdrop_bootstrap_configuration()` is called before the connection to the database is initialized. (See [`backdrop_bootstrap()`](https://docs.backdropcms.org/api/backdrop/core%21includes%21bootstrap.inc/function/backdrop_bootstrap/1).)


```php
case BACKDROP_BOOTSTRAP_CONFIGURATION:
  _backdrop_bootstrap_configuration();
  break;

case BACKDROP_BOOTSTRAP_PAGE_CACHE:
  _backdrop_bootstrap_page_cache();
  break;

case BACKDROP_BOOTSTRAP_DATABASE:
  _backdrop_bootstrap_database();
  break;
```

**Comment by kiamlaluno** [comment]
BackdropDatabaseCache, the class that stores cache data in the database, uses the following code in [`BackdropDatabaseCache::__construct()`](https://docs.backdropcms.org/api/backdrop/core%21includes%21cache.inc/function/BackdropDatabaseCache%3A%3A__construct/1).

```php
// Bootstrap the database if it is not yet available.
if (!function_exists('db_query') || backdrop_get_bootstrap_phase() < BACKDROP_BOOTSTRAP_DATABASE) {
  backdrop_bootstrap(BACKDROP_BOOTSTRAP_DATABASE, FALSE);
}
```

Similar code should be used in this case.

**Comment by kiamlaluno** [comment]
I think the following code is not correct.

```php
preg_match('/^db:(\/(\w*)\/)?(\w+)$/', trim($dburl), $matches);
if (count($matches) != 4) {
  throw new ConfigStorageException(t('Invalid database specifier: @db', array('@db' => $dburl)));
}
```

`count($matches)` could also be 2 or 3.

**Comment by laryn** [comment]
@kiamlaluno I did a test in regard to your last comment and I think the code is good (the `preg_match` returns an empty result that is counted in the total for a few of those capturing groups):

![image](https://github.com/backdrop/backdrop-issues/assets/871421/c303c1b7-7f88-4b46-9426-bfeeaa2283ad)

@herbdool and @kiamlaluno -- thanks to both of you for the feedback. I've made some revisions in the PR. Care to test again?

**Comment by kiamlaluno** [comment]
My bad: I checked with code that removed empty strings from `$matches`. (I should have tested the regular expression directly.)

**Comment by kiamlaluno** [comment]
I was going to suggest a change, but then noticed the same change should be done for the `ConfigFileStorage` class. It is better to leave it as a follow-up issue.

**Comment by herbdool** [comment]
@laryn getting closer. I suggested a small change to reduce duplication.

I noticed a consistent test failure: "Re-created the .htaccess file in the 'active' config directory." fails. Which is curious since I'm assuming the tests are still running with the active config stored in files.

**Comment by kiamlaluno** [comment]
That test checks the *.htaccess* file exists for the active configuration directory, it deletes it, visits admin/reports/status, and verifies the the *.htaccess* file is recreated. That file should be always created and the test should verify it is effectively created. (I still have to understand which function is going to create that file, though.)

There should be tests that verify what happens when the storage is switched.

**Comment by kiamlaluno** [comment]
(The function that creates a .htaccess file in the configuration directory is [`system_requirements()`](https://docs.backdropcms.org/api/backdrop/core%21modules%21system%21system.install/function/system_requirements/1).)

**Comment by laryn** [comment]
I've removed the duplication and rebased.

I also haven't figured why the "Re-created the .htaccess file in the 'active' config directory." fails.

Regarding new tests, this PR doesn't support switching back and forth, it just opens the functionality. I think we'll need follow up(s) to provide a way to switch back and forth (and test that) if we want to allow that. But maybe an initial test to make sure a fresh install with database storage works would be good on this PR.

**Comment by kiamlaluno** [comment]
Just to make my previous comment clearer: I do expect that changing the used config storage and returning to the previous one, the new config values are passed from a config storage to the other one.

**Comment by laryn** [comment]
I can see a case for allowing switching, though hopefully people wouldn't do it often. The button would probably need to put the site in maintenance mode and trigger the copying of all the configuration from one storage to the other, then take it out of maintenance mode again.

I just tested the latest and see this during the installation of a ConfigDatabaseStorage so I'm making a new commit for the PR:

![image](https://github.com/backdrop/backdrop-issues/assets/871421/59992e26-88b0-49d0-abf6-aa54c92ba8a1)

**Comment by laryn** [comment]
Ready for another review on the recent changes. Looks like the latest commit fixed that broken test somehow :+1: 

If anyone is able to help with the new test for DB storage please do, otherwise I'll get to it (maybe next week).

**Comment by argiepiano** [comment]
I'm looking at the Files Changed tab for the PR and I think something went wrong with the rebase. It's showing tons of changes (in 147 files!) most of which seem unrelated. I'm wondering if the rebase you did was actually from the current 1.x version? There are many changes in js files (which are probably those for the jquery work). This makes it extremely hard to do code reviews. Any chance you can take a look?

**Comment by laryn** [comment]
Hmm.. Maybe I did do something wrong. But I'm pretty sure we need all the unrelated changes (e.g. for Tugboat previews to be rebuilt: https://github.com/backdrop/backdrop/pull/4453#issuecomment-1642415490)

**Comment by argiepiano** [comment]
OK, maybe I don't know enough about rebasing, but what I understand is that you rebase the version of Backdrop that your fork currently uses, which should match the one in the 1.x branch of Backdrop. If this is the case, then the PR should not show any of the unrelated changes, since those should already be part of the 1.x branch. What I think may have happened is that perhaps you are working with a fork of a branch that's not Backdrop's 1.x (perhaps 1.25.x?). I believe @quicksketch merged those jQuery PRs only in branch 1.x yesterday (not 1.25.x).

**Comment by laryn** [comment]
I may have merged `1.x` instead of rebasing. Now to figure out how to fix my mess, but I'm out of time for today...

**Comment by kiamlaluno** [comment]
When the branch I use for the PR is behind the source branch, I always use *Sync fork* from GitHub. When I do so, only the files I changed for the PR are shown as changed.  
See https://github.com/backdrop/backdrop/pull/4408, for example. It does not show as changed files changed by other PRs.

**Comment by argiepiano** [comment]
It looks like you committed the jquery and the php 8.2 PRs to your own repo, instead of rebasing your fork. See:
https://github.com/laryn/backdrop/commit/802ab4ced8745a4d9104010f0a5de19654aadc17
and
https://github.com/laryn/backdrop/commit/78b6f500a9ac891726d54ba4dcef6bf8dfe4ef81

And your fork still shows it's 11 commits behind 1.x. See:
https://github.com/laryn/backdrop

So, yes, it looks like this was not rebased properly.

**Comment by laryn** [comment]
@argiepiano The branch in question has been synced:

 - https://github.com/laryn/backdrop/tree/1.x-issue-2277

I believe I did sync the fork before making changes, but muddled things up by merging the changes since then manually, instead of syncing again.

**Comment by ericfoy** [comment]
A possibly spurious comment from a guy on the fringes of this issue, for what it's worth. I bring this up only because I haven't seen it mentioned (I might have missed it, though)...

It is important for someone (like me) who doesn't know exactly why I would want my site configured either way, that it be *easy to switch from one mode to the other.*

That's all.

**Comment by herbdool** [comment]
@ericfoy I suggest making a new issue: "Easily switch from storing config in files to db" instead of blocking this issue. Having the ability to store config in the db is good enough for the majority of uses, which itself is an edge case to help deal with hosting on Pantheon (and others?). Most people will have no reason to easily switch between one and the other.

**Comment by herbdool** [comment]
> That test checks the .htaccess file exists for the active configuration directory, it deletes it, visits admin/reports/status, and verifies the the .htaccess file is recreated. That file should be always created and the test should verify it is effectively created. (I still have to understand which function is going to create that file, though.)

I'm still getting this error when running tests locally. I've got the PR but I'm initially testing with the default of storing config in files:

> file_get_contents(./files/simpletest/497215/config_active/.htaccess): Failed to open stream: No such file or directory

Somewhat related to that, it seems that that section of `system_requirements()` assumes that the active and staging both store in the same place. But people can put the active in db and staging in file, so it should be checking them individually.

**Comment by laryn** [comment]
I ran stuck here trying to figure out how to write tests for this. Anyone want to carry the torch over the finish line? 🙏

**Comment by kiamlaluno** [comment]
I think the following test method should be rewritten.

```php
  /**
   * Tests that config directories are protected by .htaccess files.
   */
  public function testConfigHtaccess() {
    // Check that the .htaccess file is present in config location.
    $active_dir = config_get_config_directory();
    $this->assertTrue(is_file($active_dir . '/.htaccess'), "Located the .htaccess file in the 'active' config directory.");

    // Visit the status report to confirm the file is re-created automatically.
    $admin_user = $this->backdropCreateUser(array(
      'access administration pages',
      'access site reports',
    ));
    $this->backdropLogin($admin_user);

    // Remove .htaccess file so we can test that it gets re-created.
    @backdrop_unlink($active_dir . '/.htaccess');
    $this->assertFalse(is_file($active_dir . '/.htaccess'), "Removed the .htaccess file in the 'active' config directory.");

    $this->backdropGet('admin/reports/status');
    $this->assertTrue(is_file($active_dir . '/.htaccess'), "Re-created the .htaccess file in the 'active' config directory.");

    // Verify contents of .htaccess file.
    $file = file_get_contents($active_dir . '/.htaccess');
    $this->assertEqual($file, file_htaccess_lines(), 'The .htaccess file contains the proper content.');
  }
}
```

As it is, it checks the _.htaccess_ file is always present and that it is recreated when admin/reports/status is visited.

The configuration storage class should decide what needs to be done in that case. When the storage class uses files, it should create the _.htaccess_ file in the directory used to store the configuration files; when the storage class uses a database table, it should create that table, if it does not exist.

**Comment by kiamlaluno** [comment]
The following code would be simpler, if throwing an exception would be left to the storage class.

```php
      if (is_a($config_storage, 'ConfigFileStorage')) {
        $directory = config_get_config_directory('active');
        $exception_message = "The configuration directory in settings.php is specified as '$directory', but this directory is either empty or missing crucial files. Check that the \$config_directories variable is correct in settings.php.";
      }
      else {
        $exception_message = "The active configuration location is either empty or missing crucial information. Check that the \$settings['config_active_class'] variable is correct in settings.php.";
      }
```

**Comment by laryn** [comment]
I'm not sure what the chances are we can get this finished up in time for the next feature freeze...

Here's my summary of what remains, from what I can see:

- @herbdool comments above ([link](https://github.com/backdrop/backdrop-issues/issues/2277#issuecomment-1710599001))
- Tests. @kiamlaluno has a few comments just above, and this is from @jlfranklin on Zulip regarding writing tests for this: "I think the easiest way would be to create an extra test pipeline that uses some non-default config backend and run a subset of the tests against it. Off the cuff, I'd think the Bootstrap, Configuration, Update, and Upgrade Path test groups would be a good start. They all test config directly, early stage or out-of-band code paths."

**Comment by laryn** [comment]
@hosef I removed myself and added you based on the dev call today. I hope that's okay. :)

**Comment by quicksketch** [comment]
To help shorten the PR to only relevant changes, I separated out the code style fixes to https://github.com/backdrop/backdrop/pull/4605, which should have no functional difference. Once that's merged we can update https://github.com/backdrop/backdrop/pull/4453 with 1.x.

**Comment by quicksketch** [comment]
Updating https://github.com/backdrop/backdrop/pull/4453 with the latest 1.x resulted in dozens of merge conflicts. So I started fresh with a new PR at https://github.com/backdrop/backdrop/pull/4606

**Comment by quicksketch** [comment]
I'm doing some experimentation with getting tests running over in https://github.com/backdrop/backdrop/pull/4608. I split out the redundant parts of setting up a webserver into a separate action file, then run the normal set of tests on file config, and a subset of tests on the database config (and only on PHP 8.3). So far the traditional tests are (mostly) passing; PHP 5.6 apparently needs special treatment but I'm not sure what. All the tests run with DB config are failing. I think there's further updating needed to simpletest module to get it running alternate forms of config. Also the `--cache` parameter on running tests does not work, but probably related to all tests failing. Backdrop _does_ at least install, so that's something.

**Comment by quicksketch** [comment]
I've made a little more progress: tests can now be run using DB config, but only if not using cached profiles. Running tests locally works great, both from the command line and the UI. In GitHub Actions there are still the multiple problems listed above.

**Comment by hosef** [comment]
I was messing around with the code from your PR, and I was wondering if it would be better to static cache the storage controllers in `config_get_config_storage()`. The way it is now, each time that is called, it creates a new object. If we ensured that there was only 1 controller object for each config location, we could make some optimizations to reduce the number of times we queried the same data.

**Comment by hosef** [comment]
As far as performance goes, it does seem that database storage is slower than file storage when they are both local. The difference is only a few ms in most cases, so probably no one will notice a difference unless your environment is skewed heavily in favor of one or the other.

**Comment by quicksketch** [comment]
Super, thanks for trying it out and benchmarking @hosef!

> I was wondering if it would be better to static cache the storage controllers in `config_get_config_storage()`

That might also be worth benchmarking; though I don't think that's the focus of this PR. The constructor for creating a config object is not very complicated so there may not be a noticeable benefit.

**Comment by hosef** [comment]
Initially it would not make a difference. The main benefit would be the ability to store data in the controller about what has already been loaded, so that we don't need to reload data multiple times.

**Comment by laryn** [comment]
Thanks for the work on this @quicksketch, and for looking it over @hosef!
Would it be safe to milestone this one for `1.28.0`?

**Comment by quicksketch** [comment]
Yes, let's shoot fro 1.28.0! Would love to see this. I'll see if I can help further tests as well, but I don't have any timeline for it. Anyone can try to pick up the work from https://github.com/backdrop/backdrop/pull/4608

**Comment by hosef** [comment]
@quicksketch Do I need to apply both of your PRs to run tests for this?

**Comment by quicksketch** [comment]
@hosef No, just https://github.com/backdrop/backdrop/pull/4608 where the most progress is made. It includes all the changes from https://github.com/backdrop/backdrop/pull/4606, which was just a reroll of the previous PR (and did not attempt to include any test coverage yet).

**Comment by laryn** [comment]
@quicksketch Were you saying you prefer to change `db:/` to `db://`? Should we just do that now?

**Comment by quicksketch** [comment]
Yes, I think that's the minimal change that would make me feel a lot better about the connection format. If anyone has any other ideas about a nicer format that would be appreciated. Right now this isn't publicly exposed, but I could see this being a settings.php option that allowed using a different database connection, for example.

**Comment by quicksketch** [comment]
@hosef I pulled your PR into https://github.com/backdrop/backdrop/pull/4606 and then cherry-picked it into https://github.com/backdrop/backdrop/pull/4608 (resolving the conflicts). Tests can run locally but they're still failing when run by CI. Also seems like something broke the PHP 5.6 tests with this change.

**Comment by hosef** [comment]
@quicksketch I did a test where I switched the test runner back to using the core test script and PHP 5.6 started working again. I think there may be something missing that we need to add to `.github/actions/set-up-webserver/action.yml`.

I did another test changing all the tests to use db config which broke a lot of tests.

**Comment by hosef** [comment]
> I did another test changing all the tests to use db config which broke a lot of tests.

It looks like the tests that broke are using BackdropUnitTestCase, but are assuming that Config is available even when the database is not. These tests likely cannot be run with db config unless we change all of them to use BackdropWebTestCase.

**Comment by laryn** [comment]
I ran into this on a test site with this patch installed, when trying to sync config:

> Error: Call to undefined function backdrop_json_encode() in ConfigDatabaseStorage->encode() (line 1518 of /code/core/includes/config.inc).

I notice that in `->decode()` we're using `json_decode` directly, for efficiency. Perhaps we should use `json_encode` directly in `->encode()`? 

I am testing that locally. I ran into an issue next which had to do with XAutoload. I was removing the configuration for the module although the module was still enabled. (Oops). But it either saved the configuration with a null value, or that was regenerated automatically by the module somehow. I had to insert default config for xautoload back into the `config_active` table to get the test site loading again.

**Comment by laryn** [comment]
> It looks like the tests that broke are using BackdropUnitTestCase, but are assuming that Config is available even when the database is not. These tests likely cannot be run with db config unless we change all of them to use BackdropWebTestCase.

Would it be acceptable to leave these tests as standard file config, and then "switch over" to testing db config after they've run?

**Comment by hosef** [comment]
We were planning on just running the config tests with the database config backend. I was doing that mostly to see if it would give us any new information about what is going on.

**Comment by quicksketch** [comment]
I updated the PR at https://github.com/backdrop/backdrop/pull/4608 with a few small changes:

- Changed the syntax for a config connection from `db:/[db_name]/[table]` to `db://[db_name]/[table]` (adding an additional slash).
- Renamed the default config table names from `active_config` and `staging_config` to `config_active` and `config_staging`.
- Renamed the table column name from `ctime` to `modified`, a timestamp value. This makes it match the `node` table and others for tracking last modified time.

I tested I'm still able to install Backdrop with the new config system and run all config tests. I also tested running a number of Unit tests, as @hosef mentioned some of them assume a file-based config system. I couldn't reproduce a problem though; @hosef are there any particular unit tests you know of that have a problem? Or maybe it only occurs when both active and staging are stored in the database? I've only tried with `active` in the database.

**Comment by quicksketch** [comment]
I enabled artifacts to upload the Simpletest Verbose output to the test runs (I also filed a request to do this on all runs: https://github.com/backdrop/backdrop-issues/issues/6458), and discovered that the tests aren't passing for an obvious reason: Apache doesn't seem to be set up correctly in our GitHub Action:
![image](https://github.com/backdrop/backdrop-issues/assets/100206/85993c64-f5b7-4d4c-bbcf-3d56397b8870)

So if we can figure that out, tests should mostly start working. PHP 5.3 tests are also failing for some reason but this approach of attaching the verbose output might help with it too.

**Comment by quicksketch** [comment]
Sigh. We've been stuck for weeks on this it was caused by a stray space in the Apache setup code:

```diff
-     sudo sed -i -e "s/_PHP_VERSION_/ ${{ inputs.php-version }}/" /etc/apache2/sites-available/000-default.conf
+     sudo sed -i -e "s/_PHP_VERSION_/${{ inputs.php-version }}/" /etc/apache2/sites-available/000-default.conf
```

It's crazy that any of the tests were passing at all with this typo. But hey, now config tests are passing. :tada:

![image](https://github.com/backdrop/backdrop-issues/assets/100206/efefea1e-f0cf-47c6-a2b5-57b54db442b6)

Still looking into the PHP 5.3 failures.

**Comment by herbdool** [comment]
The small syntax fixes, that eat up hours of your time debugging, are both annoying and a relief.

**Comment by quicksketch** [comment]
I looked at this again today and all tests are now passing except for a gnarly issue with `ImageStylesPathAndUrlUnitTest`, which I don't understand how it's even being affected by these changes. It's also passing locally for me, and the verbose output trick I used above (see https://github.com/backdrop/backdrop-issues/issues/6458) didn't help because the verbose output is correct between GitHub Actions and my local. It's only the Content-Length header that is wrong and I can't figure out why.

**Comment by quicksketch** [comment]
The issue with `ImageStylesPathAndUrlUnitTest` seems to be affecting all current pull requests, see https://github.com/backdrop/backdrop-issues/issues/6459.

I'll rollback my changes in trying to fix that issue, it should be handled separately. The good news is that this means that all expected tests are passing.

**Comment by quicksketch** [comment]
The `ImageStylesPathAndUrlUnitTest` issue aside, this is now passing all tests and is ready for review!

https://github.com/backdrop/backdrop/pull/4608

**Comment by indigoxela** [comment]
There's still some test nagging, but I finally got curious, how this all would work in B (I don't like it the way it's done in D10).

It works!

With a fresh - yet not installed - Backdrop with patch applied I set
```
...
$settings['config_active_class'] = 'ConfigDatabaseStorage';
$config_directories['staging'] = '../bd-config/staging';
...
```
... in settings.local.php, then installed the site.

Verified that there's a "config_active" database table (yay, data is longtext, no binary stuff :heart_eyes:).
Then went to /admin/config/development/configuration/full/export and downloaded a full export.
Went to /admin/config/system/site-information and made some changes (checked the changes).
Did a full import of my downloaded file on /admin/config/development/configuration/full/import, reviewed the staged config diff, confirmed and applied the backup. And the settings were back.

Everything's fast and works. :+1: 

A bit of a surprise: the db table column "changed" doesn't seem to get updated on config changes? :disappointed: 

However, went on and installed a contrib module (Tinymce), that worked fine (config was added). Then added a custom editor profile (which added another config row in db), then edited that profile, which worked, but also didn't update the "changed" column in the db table. Hm. Either I misunderstood something, or this is really not working properly.

**Comment by herbdool** [comment]
I see a write method which updates the `changed` column to `time()` on each write to config so I'm not sure why the doesn't appear to be changing sometimes.

**Comment by quicksketch** [comment]
> A bit of a surprise: the db table column "changed" doesn't seem to get updated on config changes?

Nice catch @indigoxela! You are right. For some reason "changed" was only getting updated on create, not on updates.This is fixed now.

I also found that if the database is used for _staging_ config, that it was not importing the full archive correctly. This is fixed too.

There's one test intermittently failing (`MenuTrailTestCase`) that I'm trying to sort out.

@indigoxela If you have the space to review the re-arrangement of the GitHub Action code that would be great. I abstracted the webserver setup into a dedicated action file that can be shared by the two SimpleTest workflows (one for file-based config and one for db-based config).

**Comment by quicksketch** [comment]
I figured out the intermittent `MenuTrailTestCase` failure and tests are passing consistently now.

**Comment by indigoxela** [comment]
>  If you have the space to review the re-arrangement of the GitHub Action code...

Done. :wink:

Currently the "Database Config" runs on a single PHP version - and needs another runner. Excluding the Tugboat job, we're up to 12 runners now (the limit for us is 20 at once). If we add all php versions, our runner usage per PR would be 15... It's already the case that we can not run two PR checks at a time. We should keep an eye on this increasing jobs number...

**Comment by indigoxela** [comment]
I'd like to note that the current PR breaks almost every [Bee](https://github.com/backdrop-contrib/bee) command, if the site's config class is ConfigDatabaseStorage. And it doesn't seem like Bee does anything "dirty" with config.

**Comment by indigoxela** [comment]
> This will probably break contrib as well - thinking of backup_migrate.

I did a quick check: yes, B&M is also affected, though in a slightly different way. :disappointed:

**Comment by quicksketch** [comment]
> I'd like to note that the current PR breaks almost every [Bee](https://github.com/backdrop-contrib/bee) command, if the site's config class is `ConfigDatabaseStorage`. And it doesn't seem like Bee does anything "dirty" with config. Eventually a bootstrap level problem?

I wasn't able to reproduce a problem with `bee`. Running [bee in DDEV](https://github.com/backdrop-contrib/bee/wiki/Using-bee-with-DDEV) my commands all seem to work fine, even the ones directly manipulating config:

![image](https://github.com/backdrop/backdrop-issues/assets/100206/fc38d4cd-fd1e-4610-9434-5717f0cca372)

All of the following seem to have worked for me:

```
bee uli
bee state-get install_time
bee config-get system.core site_name
bee config-set system.core site_name New-Value
```

And I confirmed that the active config is being stored in the database by querying and seeing my new site_name key being set.
```
select * from config_active ORDER BY changed DESC LIMIT 1;
```

As for B&M, I would expect it will need to take into account the Database config as a feature improvement. As long as this doesn't break the existing file-based config storage, I don't think that should be a blocker.

@indigoxela I read over your GHA review (thank you!) and it doesn't look like there are any action items. I'm going to put this back to review. Please let me know if I missed something that needs attention.

**Comment by indigoxela** [comment]
@quicksketch re Bee - could you provide some details about your setup? I can't seem to get any of the commands you listed working (only "bee st" works). I just get "Exception: The configuration directory... active... does not exist. in config_get_config_directory() (line 235 of .../core/includes/config.inc)". (With latest bee via git, with PHP 8.2)

My relevant settings:
```
$settings['config_active_class'] = 'ConfigDatabaseStorage';
$config_directories['active'] = NULL;
$config_directories['staging'] = '../bd-config/staging';
```
I tried combinations with different "active" dir settings, or commenting out. No clue, how this could work.

Is there some leftover active config dir in your install? Maybe from earlier installs, inactive now, but existing, that just makes the check in _backdrop_bootstrap_configuration() pass?

**Comment by laryn** [comment]
I have been testing locally on a copy of a live site that I've switched over to config in the database. I can confirm the issue with Bee, but everything else I've tested is working very well (changing configuration via the UI, importing/exporting configuration one-off and globally).

**Comment by laryn** [comment]
It seems as though this check is failing and sending back the default when bee is used:
```
function config_get_config_storage($type = 'active') {
  $class = settings_get('config_' . $type . '_class', 'ConfigFileStorage');
```

**Comment by laryn** [comment]
When running `bee cc all` It looks like `backdrop_settings_initialize()` first gets called from `bee_initialize_console()` and the values of `$settings` are set correctly, including `config_active_class`.

Then `bee_process_command()` calls `bee_bootstrap()` and that in turn calls `backdrop_bootstrap()` and we end up in `backdrop_settings_initialize()` again. In the spot where settings gets initialized to an empty array, the previously loaded values are zero'ed out, and this time around they are not reset, because this condition is not met:
`if (isset($_SERVER['BACKDROP_SETTINGS'])) {`

I don't know why that setting is no longer set!

**Comment by quicksketch** [comment]
Thanks @indigoxela, you guessed correctly. I had left-over files that made it so the config files were being found even though they weren't the ones actually being used.

I remember we had this problem with `bee` previously, here's the issue: https://github.com/backdrop-contrib/bee/issues/205#issuecomment-1310821674

`bee` worked-around the problem without patching core by duplicating `$settings` temporarily and then restoring it.

Now seems like a good time to at least try fixing this in core, I think it might be as simple as not wiping out the `$settings` array in `backdrop_settings_initialize()`; that would allow it to be called without loosing the existing settings values. I'll update the PR with this approach and see what happens.

**Comment by quicksketch** [comment]
By [removing a single line](https://github.com/backdrop/backdrop/pull/4608/commits/5e2043eb7c50d6424b7ee359cb5b208303f2ff7d) that reset the global `$settings`, it has both fixed `bee` and still has 100% test coverage passing.

**Comment by laryn** [comment]
Do we have any core committers who haven’t been involved in this issue in some way? It’s looking great to me…

**Comment by quicksketch** [comment]
Looks like after merging in 1.x, we have a few different failures. They're all similar in that the active config directory can't be found while setting up new tests. It seems to happen randomly based on the failures listed. So there's something that still needs addressing in the test setup is seems.

**Comment by quicksketch** [comment]
I put in a new fix for the tests, the failure in `MenuTrailTestCase` is a weird one where a test can fail if the previous test has unwritten cache entries that need to be saved into the theme registry. The cache tries to write in-between test cases when there isn't an active config directory. I fixed it by adding a `backdrop_static_reset()` into `BackdropWebTestCase::tearDown()`, which makes the theme registry write its pending cache entries. This seems to have fully fixed the problem and I got 3 full test runs with 100% passing.

**Comment by laryn** [comment]
7.5 years, 140 comments, 4 pull requests, many hours of puzzlement later, merged into 1.x as a new feature to be included in `1.28.x`!

Big thanks to @sentaidigital and @jlfranklin for starting things off, and to @quicksketch and @hosef for the heavy lifting of getting tests working at the end. With reviews and assists in the middle from @herbdool, @laryn, @indigoxela, @mikemccaffrey, @jenlampton, @serundeputy, @Gormartsen, @klonos, @alexfinnarn, @kiamlaluno, @argiepiano. 

A great team effort! 🎉🎉🎉

Reopening for the needed change record.

**Comment by jenlampton** [comment]
Reopening for Change Record
